### PR TITLE
Fix/geo fixes

### PR DESCRIPTION
**Pull request by domthomhive**
Fixes a number of geolocation issues found across a few different android versions:

- Geofence broadcast was incorrectly defined and was trying to start an activity rather than broadcasting a geofence event
- No error handling in place when the settings check failed (was returning error code 7 with no resolution)
- Some missing intent filters to re-start geofence service when location providers changed
- Newer version of android were trying to get location updates rather than just retrieving the current location. This was causing delays/timeouts and was not consistently returning location data each time it was accessed.
- Gradle wrapper/kotlin was not configured in project so code wasn't compiling making development impossible.
- Missing intent filter for when location providers changed. This needs the app to re-start the geofence event broadcast receiver

### CODE REVIEW

File: android/src/main/java/co/uk/hive/reactnativegeolocation/location/LocationController.java [hunk 1]
```diff
@@ -116,73 +113,43 @@ private boolean hasPermissions() {
...
 
         if (ActivityCompat.checkSelfPermission(mContext, Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED
                 && ActivityCompat.checkSelfPermission(mContext, Manifest.permission.ACCESS_COARSE_LOCATION) != PackageManager.PERMISSION_GRANTED) {
             final String errorMessage = "Missing the required location permissions!";
             Log.e(LocationController.class.getName(), errorMessage);
-            failureCallback.apply(errorMessage);
+            failureCallback.apply(LocationError.PERMISSION_DENIED);
             return;
         }
-        mLocationClient.requestLocationUpdates(getLocationRequest(currentPositionRequest), locationCallback, Looper.getMainLooper());
```
Comment (domthomhive) [inline android/src/main/java/co/uk/hive/reactnativegeolocation/location/LocationController.java:151]: Requesting location updates was unreliable and wouldn't always complete. Assuming there was some throttling happening from the OS and the app was often stuck and timing out when trying to access the current location.

File: android/src/main/java/co/uk/hive/reactnativegeolocation/RNGeolocationModule.java [hunk 2]
```diff
@@ -87,29 +116,82 @@ public void isLocationEnabled(Promise promise) {
...
         };
     }
+
+    private void handleApiException(ApiException apiException, Callback failureCallback) {
+        switch (apiException.getStatusCode()) {
+            case LocationSettingsStatusCodes.RESOLUTION_REQUIRED: {
+                // Location settings are not satisfied. But could be fixed by showing the user a dialog.
+                try {
+                    // Cast to a resolvable exception.
+                    final ResolvableApiException resolvable = (ResolvableApiException) apiException;
```
Comment (domthomhive) [inline android/src/main/java/co/uk/hive/reactnativegeolocation/RNGeolocationModule.java:140]: For Android 10 and above, if the settings check fails, this checks to see if the the issue is resolvable and presents a dialog allowing the user to enable high accuracy directly..

File: android/src/main/AndroidManifest.xml [hunk 3]
```diff
@@ -28,7 +29,10 @@
             <intent-filter>
                 <action android:name="android.intent.action.BOOT_COMPLETED"/>
                 <action android:name="android.intent.action.MY_PACKAGE_REPLACED"/>
+                <action android:name="android.intent.action.QUICKBOOT_POWERON" />
+                <action android:name="com.htc.intent.action.QUICKBOOT_POWERON" />
                 <action android:name="android.location.MODE_CHANGED"/>
+                <action android:name="android.location.PROVIDERS_CHANGED" />
```
Comment (domthomhive) [inline android/src/main/AndroidManifest.xml:35]: Here we add some more system events that require us to restart the geofence broadcast receiver.

File: android/src/main/java/co/uk/hive/reactnativegeolocation/location/LocationController.java [hunk 4]
```diff
@@ -75,39 +76,35 @@ public void getCurrentPosition(CurrentPositionRequest currentPositionRequest,
...
-                    .setPriority(Priority.PRIORITY_HIGH_ACCURACY)
-                    .setGranularity(Granularity.GRANULARITY_FINE)
-                    .setDurationMillis(CURRENT_LOCATION_REQUEST_DURATION_MILLIS).build();
-
-            final Task<Location> currentLocationTask = mLocationClient.getCurrentLocation(currentLocationRequest, null);
-            currentLocationTask.addOnSuccessListener(location -> {
-                if (location == null || Double.isNaN(location.getLatitude()) || Double.isNaN(location.getLongitude())) {
-                    failureCallback.apply(LocationError.LOCATION_IS_NULL);
+            try {
+                if (getLocationMode(mContext) == Settings.Secure.LOCATION_MODE_HIGH_ACCURACY) {
```
Comment (domthomhive) [inline android/src/main/java/co/uk/hive/reactnativegeolocation/location/LocationController.java:82]: This checks if location mode is set to high accuracy on android 8 and below

File: android/src/main/java/co/uk/hive/reactnativegeolocation/location/LocationController.java [hunk 5]
```diff
@@ -116,73 +113,43 @@ private boolean hasPermissions() {
     }
 
     private LocationRequest getLocationRequest(CurrentPositionRequest currentPositionRequest) {
-        return LocationRequest.create()
-                .setNumUpdates(1)
-                .setPriority(Priority.PRIORITY_BALANCED_POWER_ACCURACY)
-                .setExpirationDuration(currentPositionRequest.getTimeout());
+        return new LocationRequest.Builder(5000)
+                .setMaxUpdates(1)
+                .setPriority(Priority.PRIORITY_HIGH_ACCURACY)
```
Comment (domthomhive) [inline android/src/main/java/co/uk/hive/reactnativegeolocation/location/LocationController.java:118]: Changing this to high accuracy priority as we don't need to monitor the location, we just need to access it once when configuring geofences or changing the home location. We want to be as accurate as possible.

**Review by bvic23 — APPROVED** [review]

**Pull request merged**